File: sdk/flutter/lib/eyex.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

class EyeXPalette {
  final String background, surface, text, primary, secondary, error, success;
  const EyeXPalette({required this.background,required this.surface,required this.text,required this.primary,required this.secondary,required this.error,required this.success});
  factory EyeXPalette.fromJson(Map<String,dynamic> j)=>EyeXPalette(background:j['background'],surface:j['surface'],text:j['text'],primary:j['primary'],secondary:j['secondary'],error:j['error'],success:j['success']);
  Map<String,dynamic> toJson()=>{'background':background,'surface':surface,'text':text,'primary':primary,'secondary':secondary,'error':error,'success':success};
}
class EyeXTheme {
  final String type; final EyeXPalette palette; final bool contrastOk;
  const EyeXTheme({required this.type,required this.palette,required this.contrastOk});
  factory EyeXTheme.fromJson(Map<String,dynamic> j)=>EyeXTheme(type:j['type'],palette:EyeXPalette.fromJson(j['palette']),contrastOk:j['contrast_ok'] as bool);
}
class EyeXTestAnswers {
  final bool redsLookDarker, greenBrownConfusion, blueYellowConfusion, colorsLookGray;
  const EyeXTestAnswers({required this.redsLookDarker,required this.greenBrownConfusion,required this.blueYellowConfusion,required this.colorsLookGray});
  Map<String,dynamic> toJson()=>{'reds_look_darker':redsLookDarker,'green_brown_confusion':greenBrownConfusion,'blue_yellow_confusion':blueYellowConfusion,'colors_look_gray':colorsLookGray};
}
class EyeXTestResult {
  final String suggestedType, disclaimer;
  const EyeXTestResult(this.suggestedType,this.disclaimer);
  factory EyeXTestResult.fromJson(Map<String,dynamic> j)=>EyeXTestResult(j['suggested_type'],j['disclaimer']);
}
class EyeXClient {
  final String baseUrl;
  const EyeXClient(this.baseUrl);
  String get _base=>baseUrl.replaceFirst(RegExp(r'/$'),'');
  Future<Map<String,dynamic>> _json(Uri uri,{String method='GET',Map<String,dynamic>? body}) async {
    final headers={'Accept':'application/json',if(body!=null)'Content-Type':'application/json'};
    final response=method=='POST'?await http.post(uri,headers:headers,body:jsonEncode(body)):await http.get(uri,headers:headers);
    final data=jsonDecode(response.body) as Map<String,dynamic>;
    if(response.statusCode<200||response.statusCode>=300)throw Exception(data['message']??data['error']??'EyeX request failed');
    return data;
  }
  Future<List<String>> types() async {final data=await _json(Uri.parse('$_base/api/v1/theme/types'));return (data['types'] as List).cast<String>();}
  Future<EyeXTheme> theme(String type,{String? severity,String? mode,bool? highContrast}) async {
    final params=<String,String>{};if(severity!=null)params['severity']=severity;if(mode!=null)params['mode']=mode;if(highContrast!=null)params['high_contrast']='$highContrast';
    final uri=Uri.parse('$_base/api/v1/theme/$type').replace(queryParameters:params.isEmpty?null:params);
    return EyeXTheme.fromJson(await _json(uri));
  }
  Future<EyeXTheme> custom(String type,EyeXPalette palette,{String? severity,String? mode,bool? highContrast}) async {
    final body=<String,dynamic>{'type':type,'palette':palette.toJson(),if(severity!=null)'severity':severity,if(mode!=null)'mode':mode,if(highContrast!=null)'high_contrast':highContrast};
    return EyeXTheme.fromJson(await _json(Uri.parse('$_base/api/v1/theme/custom'),method:'POST',body:body));
  }
  Future<EyeXTestResult> suggest(EyeXTestAnswers answers) async {
    final data=await _json(Uri.parse('$_base/api/v1/test/suggest'),method:'POST',body:{'answers':answers.toJson()});
    return EyeXTestResult.fromJson(data);
  }
}

File: frontend/html-js/eyex.js
(function () {
  'use strict';
  if (window.EyeX && window.EyeX.__widgetLoaded) return;

  var script = document.currentScript;
  var src = script && script.src ? new URL(script.src, window.location.href) : new URL(window.location.href);
  var apiBase = (script && script.dataset.eyexApi ? script.dataset.eyexApi : src.origin).replace(/\/$/, '');
  var keys = ['background', 'surface', 'text', 'primary', 'secondary', 'error', 'success'];
  var labels = { normal: 'Normal', protanopia: 'Protanopia', deuteranopia: 'Deuteranopia', tritanopia: 'Tritanopia', achromatopsia: 'Acromatopsia', low_vision: 'Baja visión' };

  function setPalette(data) {
    keys.forEach(function (key) { document.documentElement.style.setProperty('--eyex-' + key, data.palette[key]); });
    document.documentElement.dataset.eyexTheme = data.type;
    document.body.style.backgroundColor = 'var(--eyex-background)';
    document.body.style.color = 'var(--eyex-text)';
  }

  async function apply(type, options) {
    options = options || {};
    var params = new URLSearchParams();
    if (options.severity) params.set('severity', options.severity);
    if (options.mode) params.set('mode', options.mode);
    if (typeof options.highContrast === 'boolean') params.set('high_contrast', String(options.highContrast));
    var response = await fetch(apiBase + '/api/v1/theme/' + encodeURIComponent(type) + (params.toString() ? '?' + params : ''), { headers: { Accept: 'application/json' } });
    var data = await response.json();
    if (!response.ok) throw new Error(data.message || data.error || 'EyeX request failed');
    setPalette(data);
    localStorage.setItem('eyex-widget-theme', data.type);
    return data;
  }

  function injectBaseStyles() {
    var style = document.createElement('style');
    style.id = 'eyex-widget-style';
    style.textContent = 'body{transition:background-color .15s,color .15s}a{color:var(--eyex-primary)!important}button,input,select,textarea{border-color:var(--eyex-primary)!important}.eyex-surface,[data-eyex-surface]{background:var(--eyex-surface)!important;color:var(--eyex-text)!important}.eyex-primary,[data-eyex-primary]{background:var(--eyex-primary)!important;color:#fff!important}#eyex-widget{position:fixed;right:18px;bottom:18px;z-index:2147483647;font-family:Arial,sans-serif}#eyex-widget button,#eyex-widget select{font:13px Arial,sans-serif}#eyex-toggle{width:48px;height:48px;border:0;border-radius:24px;background:#222831;color:#fff;box-shadow:0 2px 9px #0005;cursor:pointer;font-weight:700}#eyex-panel{display:none;position:absolute;right:0;bottom:58px;width:230px;padding:12px;background:#fff;color:#20252B;border:1px solid #8b949e;box-shadow:0 4px 18px #0004}#eyex-panel.open{display:grid;gap:9px}#eyex-panel label{display:grid;gap:4px;font-size:11px;font-weight:700}#eyex-panel select{height:32px;width:100%}#eyex-apply{padding:8px;background:#2E6DA4;color:#fff;border:0;cursor:pointer}';
    document.head.appendChild(style);
  }

  function makeWidget(types) {
    var host = document.createElement('div'); host.id = 'eyex-widget';
    var panel = document.createElement('div'); panel.id = 'eyex-panel'; panel.setAttribute('role', 'dialog'); panel.setAttribute('aria-label', 'Opciones EyeX');
    var typeLabel = document.createElement('label'); typeLabel.textContent = 'Modo'; var typeSelect = document.createElement('select');
    types.forEach(function (type) { var option = document.createElement('option'); option.value = type; option.textContent = labels[type] || type; typeSelect.appendChild(option); }); typeLabel.appendChild(typeSelect);
    var severityLabel = document.createElement('label'); severityLabel.textContent = 'Intensidad'; var severity = document.createElement('select');
    [['mild','Suave'],['moderate','Moderada'],['severe','Severa']].forEach(function (item) { var o=document.createElement('option');o.value=item[0];o.textContent=item[1];if(item[0]==='moderate')o.selected=true;severity.appendChild(o); }); severityLabel.appendChild(severity);
    var modeLabel = document.createElement('label'); modeLabel.textContent = 'Tema'; var mode = document.createElement('select');
    [['dark','Oscuro'],['light','Claro']].forEach(function (item) { var o=document.createElement('option');o.value=item[0];o.textContent=item[1];mode.appendChild(o); }); modeLabel.appendChild(mode);
    var hcLabel = document.createElement('label'); var hc = document.createElement('input'); hc.type = 'checkbox'; hcLabel.textContent = 'Alto contraste '; hcLabel.appendChild(hc);
    var applyButton = document.createElement('button'); applyButton.id = 'eyex-apply'; applyButton.type = 'button'; applyButton.textContent = 'Aplicar';
    panel.appendChild(typeLabel); panel.appendChild(severityLabel); panel.appendChild(modeLabel); panel.appendChild(hcLabel); panel.appendChild(applyButton);
    var toggle = document.createElement('button'); toggle.id = 'eyex-toggle'; toggle.type = 'button'; toggle.textContent = 'EyeX'; toggle.setAttribute('aria-label', 'Abrir opciones de accesibilidad EyeX');
    toggle.addEventListener('click', function () { panel.classList.toggle('open'); });
    applyButton.addEventListener('click', function () { apply(typeSelect.value, { severity: severity.value, mode: mode.value, highContrast: hc.checked }).then(function () { panel.classList.remove('open'); }).catch(function (error) { applyButton.textContent = error.message; }); });
    host.appendChild(panel); host.appendChild(toggle); document.body.appendChild(host);
    var saved = localStorage.getItem('eyex-widget-theme'); if (saved && types.indexOf(saved) >= 0) typeSelect.value = saved;
  }

  async function init() {
    injectBaseStyles();
    try {
      var response = await fetch(apiBase + '/api/v1/theme/types'); var data = await response.json();
      if (!response.ok) throw new Error('No se pudo cargar EyeX'); makeWidget(data.types);
    } catch (error) { console.warn('[EyeX]', error); }
  }

  window.EyeX = { __widgetLoaded: true, apply: apply, apiBase: apiBase };
  if (document.readyState === 'loading') document.addEventListener('DOMContentLoaded', init); else init();
}());
